packages/variant-theme-v9/src/ThemeGenerator.ts

```diff
@@ -0,0 +1,428 @@
+  return acc;
+}, {});
+
+export const getV8Theme = (primaryColor: string, darkMode = false) => {
+  const themeRules = themeRulesStandardCreator();
+
+  ThemeGenerator.setSlot(
+    themeRules[BaseSlots[BaseSlots.foregroundColor]],
+    darkMode ? whiteHex : '#323130',
+    darkMode,
+    false,
+    darkMode
+  );
+  ThemeGenerator.setSlot(
+    themeRules[BaseSlots[BaseSlots.backgroundColor]],
+    darkMode ? '#323130' : whiteHex,
+    darkMode,
+    false,
+    darkMode
+  );
+  ThemeGenerator.setSlot(themeRules[BaseSlots[BaseSlots.primaryColor]], primaryColor, darkMode, true, true);
+
+  const themeAsJson = ThemeGenerator.getThemeAsJson(themeRules);
+
+  const finalTheme = createTheme({
+    ...{ palette: themeAsJson },
+    isInverted: darkMode || isDark(themeRules[BaseSlots[BaseSlots.backgroundColor]].color!)
+  });
+
+  return finalTheme;
+};
+
+export function getVariantTheme(
+  brandVariants: BrandVariants,
+  theme: Theme,
+  type: VariantThemeType,
+  isInverted: boolean
+) {
+  // Need a way to detect custom neutrals in the theme
+  // Need a way to detect if the theme is light or dark
+
+  const newTheme: Theme = {
+    ...theme
+  };
+  switch (type) {
+    case VariantThemeType.Neutral:
+      // Set the background to be a light grey
+      // Shift all the other neutrals to account for the darker background
+      getNeutralVariant(newTheme, isInverted);
+      return newTheme;
+    case VariantThemeType.Soft:
+      // Set the background to be a light tint of the brand color
+      // Shift all the other neutrals to account for the darker background
+      getSoftVariant(newTheme, brandVariants, isInverted);
+      return newTheme;
+    case VariantThemeType.Strong:
+      getStrongVariant(newTheme, brandVariants, isInverted);
+      return newTheme;
+    default:
+      return newTheme;
+  }
+}
+
+function getNeutralVariant(theme: Theme, isInverted: boolean) {
+  for (const token in theme) {
+    if (token.startsWith('colorNeutralBackground') && !isTokenNameContains(token, ['Alpha'])) {
+      const num: number | undefined = getNumberFromToken(token);
+      if (!num) {
+        continue;
+      }
+      if (num >= 4 && num <= 5) {
+        if (token.endsWith('hover')) {
+          theme[token] = shift(theme[token], !isInverted, 12);
+        } else if (token.endsWith('pressed') || token.endsWith('selected')) {
+          theme[token] = shift(theme[token], !isInverted, 18);
+        }
+      } else {
+        theme[token] = shift(theme[token], !isInverted, 4);
+      }
+    } else if (
+      (token.startsWith('colorNeutralStroke') &&
+        !isTokenNameContains(token, ['Alpha', 'OnBrand', 'AccessibleSelected', 'InvertedDisabled'])) ||
+      (isTokenNameContains(token, ['Foreground']) &&
+        isTokenNameContains(token, ['Neutral']) &&
+        !isTokenNameContains(token, ['Brand'])) ||
+      isTokenNameContains(token, ['Subtle'])
+    ) {
+      theme[token] = shift(theme[token], !isInverted, 4);
```

Comment:
I think some of the code is from general pinciples and some of them is just a fix-up. can we figure out a way to distinguish them?